### DISCUSSION TITLE
Generating wires from an imported float array.

### DISCUSSION THREAD

**Discussion by tim-rastall-reveal** [jimy-byerley/pymadcad#35] [Q&A] [answered]
Hi Jimy/community

Firstly, thanks for your work on pymadcad!

Secondly, I'm quite a novice with python but have exhausted google/stack-exchange so I hope you'll forgive me if this is a trivial problem:

I have a list of 3d points that I'm pulling from a csv and parsing into a 3n float array using numpy.
With these points I'd like to generate wires to then use in creating some blended surfaces.

Could you give me some guidance on how best to achieve the points-to-wire conversion? I assume there's some way of forming a points buffer to use in generating the wire but I can find any examples where the points are imported from an external source (apart from meshes).

Many thanks.

**Comment by jimy-byerley** [comment, marked as answer]
Welcome there @tim-rastall-reveal !
Thanks for your thanks ;)
A question is only trivial when you already know the answer

## Step by step

I will suppose that you have an array of `float64`, else you should first convert it.
```python
sourcepoints = np.empty((n,3), dtype='f8')
# if it's comming in a different number type, do the following
sourcepoints = mysource.astype('f8')
```

Once you have it, you can cast it into an array of `vec3`, so madcad will know that it is points, not just a matrix. That operation won't copy anything, this is just a type reinterpretation.
```python
points = typedlist(sourcepoints, dtype=vec3)
```

*Note that creating a `typedlist` in that way doesn't check for the input type, so it can be error-prone.  you can alternatively use `mesh.numpy_to_typedlist(sourcepoints)`*

Then you can simply build a `Wire` using the conversion function
```python
mywire = wire(points)
```

## In short

So more directly you can do
```python
mywire = wire(typedlist(sourcepoints.astype('f8', copy=False), dtype=vec3))
```
Then you can generate surfaces from it
```python
# for instance
mysurface = extrusion(vec3(1,2,3), mywire) 
```

**Reply by tim-rastall-reveal** [reply]
Hey Jimy, 
Thanks for the prompt and comprehensive response.
I think I've understood and implemented your suggested approach that seems to work right up to the wire creation part but at that point the following error is returned:
> TypeError: incompatible data type for Wire creation

Here's the relevant bits of the script  (I've set the coordinates string that would normally be imported as a row from the csv file as the variable 'coords' for testing purposes - they're offset geographic coordinates btw; full precision isn't required so the rounding that occurs when they are imported as f8 is fine):

```python

#set input x,y,z coordinates as comma delimited string
coords = ('43.925597190856934,-339.16851806640625,0.971143,44.15789985656738,-339.8134765625,0.96679,43.52808952331543,-340.040283203125,0.959129,43.2957878112793,-339.39532470703125,0.963482,43.925597190856934,-339.16851806640625,0.971143')

#parse coords string into 1D float64 array
floatArray = np.fromstring(coords, dtype='f8', sep=",")
print('floatArray =',floatArray)

#cast floatArray into typedlist of vector3
points = typedlist(floatArray.astype('f8', copy=False), dtype=vec3)
print('points =',points)

```
The outputs of the print commands above are as follows:

> floatArray = [  43.92559719 -339.16851807    0.971143     44.15789986 -339.81347656
    0.96679      43.52808952 -340.0402832     0.959129     43.29578781
 -339.39532471    0.963482     43.92559719 -339.16851807    0.971143  ]

> points = typedlist([dvec3( 43.9256, -339.169, 0.971143 ), dvec3( 44.1579, -339.813, 0.96679 ), dvec3( 43.5281, -340.04, 0.959129 ), dvec3( 43.2958, -339.395, 0.963482 ), dvec3( 43.9256, -339.169, 0.971143 )])

This all looks promising up to this point but when I attempt to use 'points' as the input vectors for a wire, I get the type error referred to above. 
From the debug printout, I can see that the typed list contains dvec3 objects but I was wondering if they're still supposed to be in an array which is what the square brackets around them indicates?

Apologies if this is me missing something fundamental  and thankyou again for your help :)

**Reply by jimy-byerley** [reply]
It seems it's me missing something ^^
I didn't realized that there was still this bug in the main branch: `wire` is supporting `typedlist` on the respository branch I'm currently working on, but not on the master branch in which I forgot to add the support.

So I will correct that in the next release. In the mean time, you can call `Wire` instead of `wire` (the constructor itself), that one supports `typedlist`.
I apologize for that bug.

**Reply by jimy-byerley** [reply]
>  they're offset geographic coordinates btw; full precision isn't required so the rounding that occurs when they are imported as f8 is fine):

That's lucky, because pymadcad is not good at using a different precision than `float64`. It actually is possible, but by making slight declaration changes in the module. Which prevents us to use different precision types in madcad in the same program.

**Reply by jimy-byerley** [reply]
> From the debug printout, I can see that the typed list contains dvec3 objects but I was wondering if they're still supposed to be in an array which is what the square brackets around them indicates?

They are stored in the `typedlist` as in an array, exactly the same layout as in a numpy array. So one can be converted in the other without any copy. The brackets are here to allow copy pasting the printout to a python console ;)
The only interest of `typedlist` on top of a numpy array is that it allows `append` (numpy arrays are fixed in size) and retreiving `vec3` instances from it instead of just floats. `typedlist` is a kind of numpy array that behaves like a regular python list.

**Comment by tim-rastall-reveal** [comment]
Hi again Jimy

Firstly, that hint for using 'Wire' instead of 'wire' worked perfectly and I've now been able to move on to building some geometry.

My next challenge is to create blended surfaces between 2 wires: 

To achieve this, I've been using the 'blendpair' function on 2 wires as follows:

```python
blendpair(wireA,wireB, match='length', tangents='straight')
```
This yields some interesting results where I blend a series of pairs of 12-sided Ngons and 4 sided rectangles as you can see in the image below - the sections are appended into a single mesh that then has end-caps added to form a manifold surface.

I initially thought this issue was a result of the 'handedness' of the vertex order in the wires being inverted but when I checked, everything is right-hand-rule, so in the case of the series of 12 sided Ngons that make up the first 4 sections (from the top) of the mesh shown in the image below, the vertex order and handedness is consistent - with vert index0 on the first wire being directly above vert index0 on the wire below it and so on.

In the case of the 12 sided ngons in the object shown below, the vertices between wires are being linked as follows 0-0, 1-11, 2-10, 3-9, 4-8, 5-7, 6-6 etc

I had assumed that setting the 'match' parameter value to 'closest' would fix this but unfortunately, an error is returned from blending.py as follows: 

>  m = (pts[a] + pts[b] + pts[c] + pts[d]) /4
TypeError: list indices must be integers or slices, not glm.dvec3

Any thoughts? It feels like it could be a bug but equally, I could be miss-understanding something.

I've also attached a csv with the coordinate strings for the object shown in the image below.

I hope this makes sense and thanks for your continued support!

Tim

[AAA-079_inside.csv](https://github.com/jimy-byerley/pymadcad/files/8359464/AAA-079_inside.csv)

![image](https://user-images.githubusercontent.com/82631605/160322843-d293a735-b801-4c33-acff-f29df8c55697.png)

**Reply by jimy-byerley** [reply]
Hello 
That's a different problem it seems, I suggest we move on an other topic for this question. And I will have an answer ready.

**Reply by jimy-byerley** [reply]
I created question #37 for this ;) let's continue on that one

**Comment by Yarkane** [comment]
Hey there, I have a little (and maybe a bit dumb) question about this,

I got a list of 3D points and I want to create a spline with it. To achieve this, I simply call Wire, then use tube() with the created wire to obtain something like this :
![image](https://user-images.githubusercontent.com/45920565/194839282-112e96b0-c7a6-46f5-af91-a15b39e925fa.png)

As you can see, some edges are very rough. I wanted to know is there is a way to interpolate the curve, and add points, create "arcs"
 between the points and not straight lines.
I saw in the documentation of Wire that it seems possible :
![image](https://user-images.githubusercontent.com/45920565/194839500-54ee05de-7c09-4d4b-ad20-9fe04d7ffe0f.png)

Do you have a tip to generate these kind of wires ?

Thanks a lot ! :)

**Reply by jimy-byerley** [reply]
Hi @Yarkane !

I think this is starting a different topic, but still the answer is: Yes there is ways
- If you want to smooth particular points, you can use [`bevel()`](https://pymadcad.readthedocs.io/en/latest/reference/cut.html#madcad.cut.bevel) to rework your `Wire`. This will cut the cure around the selected points and replace this portion by an arc. 
- If you are already generating the input suite of points using madcad primitives, you can simply tune the `resolution` parameter if the primitives
- If you want to add points everywhere on the curve by interpolating your input suite of points, you can create an [`Interpolated`](https://pymadcad.readthedocs.io/en/latest/reference/primitives.html#madcad.primitives.Interpolated) primitive, then convert it into `Wire`

```python
wire(Interpolated(points, resolution=('rad', 0.1)))    
# resolution parameter can help to have something smoother out of any primitive
```

For the screenshot you saw in the documentation, this is a wire made of 3 primitives  - an arc (left) a segment (middle) and a Softened spline (right). Something like the following:
```python
wire([
    ArcThrough(A, B, C),
    Segment(C, D),
    Softened([D, E, F, G]),
    ])
```

**Reply by Yarkane** [reply]
Thanks a lot ! I use Interpolated to obtain this spline 
![image](https://user-images.githubusercontent.com/45920565/195282817-cfbfc817-9dfb-46bd-8d30-f05f4e432556.png)

That's surprising how it computes fast, I assumed it would double the execution time, but it's as fast as before.

Congrats ! 👍
